src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/IconUrlToImageCacheConverter.cs

```diff
@@ -24,38 +29,127 @@ internal class IconUrlToImageCacheConverter : IValueConverter
-        // This leads to a memory savings.
-        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
+        private static readonly RequestCachePolicy RequestCacheIfAvailable = new RequestCachePolicy(RequestCacheLevel.CacheIfAvailable);
+
+        private static readonly ErrorFloodGate ErrorFloodGate = new ErrorFloodGate();
+
+        /// <summary>
+        /// Converts IconUrl from PackageItemListViewModel to an image represented by a BitmapSource
+        /// </summary>
+        /// <remarks>
+        /// We bind to a BitmapImage instead of a Uri so that we can control the decode size, since we are displaying 32x32 images, while many of the images are 128x128 or larger.
+        /// This leads to a memory savings.
+        /// </remarks>
+        /// <param name="values">
+        /// <list type="bullet">
+        /// <item>
+        /// <description>values[0]: IconUrl that points to a URL o a local file</description>
+        /// </item>
+        /// <item>
+        /// <description>values[1]: An <c>PackageArchiveReader</c> to read from the local package for embedded icons</description>
+        /// </item>
+        /// </list>
+        /// </param>
+        /// <param name="targetType">unused</param>
+        /// <param name="parameter">A BitmapImage with the default package icon</param>
+        /// <param name="culture">unused</param>
+        /// <returns>A BitmapSource with the image</returns>
+        public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture)
         {
-            var iconUrl = value as Uri;
+            if (values == null || values.Length == 0)
+            {
+                return null;
+            }
+
+            var iconUrl = values[0] as Uri;
             var defaultPackageIcon = parameter as BitmapSource;
             if (iconUrl == null)
             {
                 return null;
             }
 
-            var cachedBitmapImage = _bitmapImageCache.Get(iconUrl.ToString()) as BitmapSource;
+            var cachedBitmapImage = BitmapImageCache.Get(iconUrl.ToString()) as BitmapSource;
             if (cachedBitmapImage != null)
             {
                 return cachedBitmapImage;
             }
 
             // Some people run on networks with internal NuGet feeds, but no access to the package images on the internet.
             // This is meant to detect that kind of case, and stop spamming the network, so the app remains responsive.
-            if (_errorFloodGate.IsOpen)
+            if (ErrorFloodGate.IsOpen)
             {
                 return defaultPackageIcon;
             }
 
             var iconBitmapImage = new BitmapImage();
             iconBitmapImage.BeginInit();
-            iconBitmapImage.UriSource = iconUrl;
 
+            BitmapSource imageResult;
+
+            // Check if the URI is an Embedded Icon URI
```

Comment:
The method name (`IsEmbeddedIconUri(…)`) is so good (kudos!) that this comment is superfluous.  Remove it.

Reply:
Removed https://github.com/NuGet/NuGet.Client/pull/3192